### ISSUE TITLE
incorrect diagnostic on irrefutable while-let with break

### ISSUE THREAD

**Issue by pabigot** [rust-lang/rust#101539]
Reviewing https://github.com/rust-lang/rust/issues/87335#issuecomment-1238970455 after finding a need for a `while-let` loop that provides access to an else pattern I crafted [this program](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=568083e3e5861a1f02b7d8434a27ccba):

```rust
fn main() {
    println!("in");
    let mut x: Result<(), &'static str> = Ok(());
    'wl: while let _ = match x {
        Ok(v) => {
            println!("match ok {:?}", v);
            x = Err("text");
        },
        Err(_) => break 'wl,
    } {
        println!("body");
    }
    println!("out");
}
```

The rustc diagnostic output is:

```
warning: irrefutable `while let` pattern
  --> src/main.rs:4:16
   |
4  |       'wl: while let _ = match x {
   |  ________________^
5  | |         Ok(v) => {
6  | |             println!("match ok {:?}", v);
7  | |             x = Err("text");
8  | |         },
9  | |         Err(_) => break 'wl,
10 | |     } {
   | |_____^
   |
   = note: `#[warn(irrefutable_let_patterns)]` on by default
   = note: this pattern will always match, so the loop will never exit
   = help: consider instead using a `loop { ... }` with a `let` inside it
```

while the application output is:
```
in
match ok ()
body
out
```

Ideally the output should have no diagnostic, because the loop clearly exits.

**Comment by compiler-errors** [comment]
Why not use a `loop` here?

**Comment by pabigot** [comment]
Yes, that's a perfectly reasonable alternative; that the diagnostic is incorrect still seems worth reporting.

**Comment by compiler-errors** [comment]
Ah, didn't read the notes below the main diagnostic. 

While I do agree that the pattern is irrefutable and we probably shouldn't change that, I think all we need to do is remove the `this pattern will always match, so the loop will never exit` note if we see a control flow operator in the body of the `while`.

**Comment by pabigot** [comment]
Agree, the diagnostic of `irrefutable` is appropriate, it's the "loop will never exit" that's incorrect.  Note that in this case the control flow operator is in a match arm expression, not the body of the `while`.
